doc/designs/quic-design/tx-packetiser.md

```diff
@@ -0,0 +1,433 @@
+as the [Stream Send Buffers](#stream-send-buffers) above.
+
+### Flow Controller and Statistics Collector
+
+To make decisions about what frames to coalesce, the packetiser relies
+on the flow controller to enforce stream and connection bandwidth limits
+[RFC 9000 4.1 Data Flow Control].
+
+```c
+/*
+ * Return the maximum amount of data that is permitted for the given stream.
+ * This includes both the stream limit and it's associated connection limit.
+ */
+size_t ossl_quic_flow_maximum_size(QUIC_STREAM *stream);
+
+/*
+ * Information the flow controller that an amount of data has been queued for
+ * sending to a stream.
+ */
+int ossl_quic_flow_sent_data(QUIC_STREAM *stream, size_t bytes);
+```
+
+### Congestion Controller
+
+Also part of the frame coalescing decision is the congestion controller
+[RFC 9002].  For MVP, this will be a _just send it_.
+
+```c
+/*
+ * Pluggable congestion controller APIs go here
+ * Extract that is required from #18018
+ */
+```
+
+### QUIC Write Record Layer
+
+Coalesced frames are passed to the QUIC record layer for encryption and sending.
+To send accumulated frames as packets to the QUIC Write Record Layer:
+
+```c
+__owur int ossl_quic_send_packets(SSL_QUIC_TX_PACKETISER *tx,
+                                  QUIC_CONNECTION *conn);
+```
+
+If _stream_ is NULL, some, none or all queued packets for the connection
+_tx_ is associated with will be queued for transmission, either now or later.
+
+If _stream_ is not NULL, queued data for this stream will be sent
+to the record layer immediately, subject to congestion and flow control
+limitations.
+
+The write record layer will need a call equivalent to `writev(2)` in order to
+avoid copying data needlessly.  Each packet is defined and sent to the write
+record layer in fragments:
+
+```c
+struct quic_buffer_st {
+    unsigned char *buffer;
+    size_t length;
+};
+
+typedef struct quic_buffer_st QUIC_BUFFER;
+
+int ossl_quic_write_bytes(QUIC_STREAM *s, STACK_OF(QUIC_BUFFER));
+```
+
+The write record layer is responsible for combining QUIC packets into datagrams.
+
+### ACK Handling and Loss Detector
+
+1. When a packet is sent, the packetiser needs to inform the ACK Manager.
+2. When a packet is ACKed, inform packetiser so it can drop sent frames.
+3. When a packet is lost, inform packetiser to create retransmission packet(s).
+4. When a packet is discarded without ACK/loss, inform packetiser to clean up.
+
+```c
+__owur int ossl_quic_packet_sent(SSL_QUIC_TX_PACKETISER *tx,
+                                 OSSL_QUIC_PACKET *packet);
+__owur int ossl_quic_packet_acked(SSL_QUIC_TX_PACKETISER *tx,
+                                  OSSL_QUIC_PACKET *packet);
+__owur int ossl_quic_packet_lost(SSL_QUIC_TX_PACKETISER *tx,
+                                 OSSL_QUIC_PACKET *packet);
+__owur int ossl_quic_packet_discarded(SSL_QUIC_TX_PACKETISER *tx,
+                                      OSSL_QUIC_PACKET *packet);
```

Comment:
Item 1 on that list already has an understood API from #18676 

```` C
int ossl_ackm_on_tx_packet(OSSL_ACKM *ackm, OSSL_ACKM_TX_PKT *pkt);
````

Reply:
It wasn't there when this was written, but it is now.

A problem with parallel design and extremely long merge times: lots of tail chasing.